include/odp/api/spec/classification.h

```diff
@@ -206,6 +213,53 @@ typedef struct odp_cls_cos_param {
 
 	/** Drop policy associated with CoS */
 	odp_cls_drop_t drop_policy;
+
+	/** Random Early Detection (RED)
+	 * Random Early Detection is enabled to initiate a drop probability
+	 * for the incoming packet when the packets in the queue/pool reaches
+	 * a specified threshold.
+	 * When RED is enabled for a particular flow then further incoming
+	 * packets are assigned a drop probability based on the size of the
+	 * pool/queue and the drop probability becomes 100% when the queue/pool
+	 * is full.
+	 * RED is logically configured in the CoS and could be implemented
+	 * in either pool or queue linked to the CoS depending on
+	 * platform capabilities. Application should make sure not to link
+	 * multiple CoS with different RED or BP configuration to the same queue
+	 * or pool.
+	 * RED is enabled when the resource limit is equal to or greater than
+	 * the maximum threshold value and is disabled when resource limit
+	 * is less than or equal to minimum threshold value. */
+
+	/** A boolean to enable RED
+	 * When true, RED is enabled and configured with RED parameters.
+	 * Otherwise, RED parameters are ignored. */
+	odp_bool_t red_enable;
+
+	/** RED parameters */
+	struct {
+		/** Maximum threshold value for RED */
+		uint32_t max_threshold;
+
+		/** Minimum threshold value for RED */
+		uint32_t min_threshold;
+	} red;
```

Comment:
Threshold value meaning and units need to be specified:  packets in queue, packets allocated from pool. What are the maximum and minimum values of these? And accuracy is implementation specific?

E.g. pool size 10 000 packet, usually values for min 8000 and max 9 500 could be. Are min 1 and max 100 allowed?

Reply:
On second thought, this number of packets threshold will not work well in our case since the packet count will depend on the size of the packets and will cause an issue. It has to be a number of segments in our case and I believe NXP has an issue with this approach since they implement RED on actual packet length ignoring packet and segment meta-data.
Considering the complexity in merging this RED across the pool and queue, I would propose to revamp the API to strip this configuration from CoS and move it to pool and queue with capability exposing whether the platform supports in Queue or in Pool.
Any thought?

Reply:
@bala-manoharan I'll add this to Monday's ARCH call agenda. Also note that Travis is still reporting numerous failures on this PR.

Reply:
@Bill-Fischofer-Linaro The build issue seems to be with pktio/dpdk.c. Anyhow for the next version of this PR I will update to the head of api-next

Travis failure log
============
pktio/dpdk.c: In function ‘dpdk_pktio_init’:
pktio/dpdk.c:1013:2: error: implicit declaration of function ‘rte_log_set_global_level’ [-Werror=implicit-function-declaration]
  rte_log_set_global_level(RTE_LOG_WARNING);
  ^
pktio/dpdk.c:1013:2: error: nested extern declaration of ‘rte_log_set_global_level’ [-Werror=nested-externs]
pktio/dpdk.c: In function ‘dpdk_pktio_term’:
pktio/dpdk.c:1068:2: error: implicit declaration of function ‘RTE_ETH_FOREACH_DEV’ [-Werror=implicit-function-declaration]
  RTE_ETH_FOREACH_DEV(port_id) {
  ^
pktio/dpdk.c:1068:2: error: nested extern declaration of ‘RTE_ETH_FOREACH_DEV’ [-Werror=nested-externs]
pktio/dpdk.c:1068:31: error: expected ‘;’ before ‘{’ token
  RTE_ETH_FOREACH_DEV(port_id) {
                               ^
pktio/dpdk.c: At top level:
pktio/dpdk.c:1056:13: error: ‘dpdk_mempool_free’ defined but not used [-Werror=unused-function]
 static void dpdk_mempool_free(struct rte_mempool *mp, void *arg ODP_UNUSED)
             ^
cc1: all warnings being treated as errors
make[1]: *** [pktio/dpdk.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory `/home/travis/build/Linaro/odp/platform/linux-generic'
make: *** [all-recursive] Error 1